### ISSUE TITLE
array_is_list() reported as impossible for array{key?: T}

### ISSUE THREAD

**Issue by zonuexe** [vimeo/psalm#11905]
### Summary

An array shape whose only keys are optional, e.g. `array{a?: string}`, includes the empty array `[]` as a valid value. Since `array_is_list([]) === true`, `array_is_list($a)` on such a shape can be `true` at runtime — but Psalm treats it as impossible (`$a is never list`), producing a false positive.

### Code

```php
<?php
/** @param array{a?: string} $a */
function f(array $a): void {
    assert(array_is_list($a));
}

f([]); // array_is_list([]) === true, so the assertion holds
```

### Actual

```
ERROR: TypeDoesNotContainType - Type array{a?: string} for $a is never list<mixed>
```

The same shows up as a dead-code / `NoValue` report if `$a` is used inside `if (array_is_list($a)) { … }`, and `@psalm-trace $a` inside that branch prints `never`. Tested with Psalm 6.16.1.

### Expected

No error: `array{a?: string}` admits `[]`, which is a list, so `array_is_list($a)` is not decidable — the assertion (and the truthy branch) are reachable.

### Notes

Found while comparing analyzers on array-shape / list semantics (PHPStan, Psalm, Mago, Phan): https://github.com/phpstan/phpstan/discussions/14939#discussioncomment-17579178. PHPStan and Mago have the same false verdict (PHPStan tracked as https://github.com/phpstan/phpstan/issues/14938); Phan reports it correctly as *maybe* a list. (Phan's `array_is_list()`-true branch narrows the shape to `list<…>`, its false branch to `non-empty-associative-array<…>`.)

**Comment by zonuexe** [comment]
Since psalm.dev is currently returning an Internal Server Error, it is being visualized using pzoom.

 * https://psalm.dev/r/36ba6550f6
 * https://pzoom.dev/#JcoxCoNAEAXQOnOKX2zhbmNvAnoT-WhMBuK6zGyEIN49kNTv3fryLCJtShgKjStoxs_BvoNX0_w4EYjUyvLOU9UtY2l-BYGxw77pjEMudL9b_cuoPr7UaxMY41XOLw
